src/utils/SaveFile.js

```diff
@@ -1,79 +1,236 @@
+    ...shareOptions,
+  };
+  return new Promise(res =>
+    window.plugins.socialsharing.shareWithOptions(options, () => res(shareURL), () => res('cancelled')));
+};
+
+const tryFileCleanup = (fileEntries) => {
+  if (fileEntries) {
+    fileEntries.map((fileEntry) => {
+      try {
+        return fileEntry.remove();
+      } catch (err) { return err; }
+    });
+  }
+};
+
+const saveZippedSessionsCordova = (exportedSessionIds, sessions, installedProtocols) => {
+  const defaultFileName = 'networkCanvasExport.zip';
+  const fileType = 'text/xml';
+  const shareOptions = { message: 'Your network canvas zip file.', subject: 'network canvas export' };
+
+  let fileSystem;
+  let dataFilenames;
+  let fileEntries;
+  return new Promise((resolve, reject) => {
+    getFileSystem()
+      .then((fs) => {
+        fileSystem = fs;
+        const promisedExports = exportedSessionIds.map((sessionId) => {
+          const session = sessions[sessionId];
+          const sessionCodebook = installedProtocols[session.protocolUID].codebook;
+
+          // export one file at a time
+          return exportCordovaFile(fs, session, sessionId, sessionCodebook, fileType);
+        });
+        // collect exported temp file names
+        return Promise.all(promisedExports);
+      })
+      .then((files) => {
+        dataFilenames = files.map(file => file.filename);
+        fileEntries = files.map(file => file.fileEntry);
+        return getFile(defaultFileName, fileSystem);
+      })
+      .then(fe => createWriter(fe))
+      .then(({ fileWriter, fileURL }) => archive(dataFilenames, fileURL, fileWriter, fileSystem))
+      .then(shareURL => cordovaShare(shareURL, shareOptions))
+      .then(() => tryFileCleanup(fileEntries))
+      .then(() => resolve(dataFilenames))
+      .catch(err => reject(err));
+  });
+};
+
+const tryUnlink = (filePath, fs) => (new Promise((resolve) => {
+  try {
+    fs.unlink(filePath, (err, ...args) => {
+      if (err) {
+        resolve(err);
+      } else {
+        resolve(...args);
+      }
+    });
+  } catch (err) { resolve(err); }
+}));
+
+const exportElectronFile = (fs, session, sessionCodebook, dataFileName) => {
+  const formatter = new GraphMLFormatter(session.network, false, false, sessionCodebook);
+
+  return new Promise((resolve, reject) => {
+    const writeStream = fs.createWriteStream(dataFileName);
+    writeStream.on('finish', () => resolve(dataFileName));
+    writeStream.on('error', err => reject(err));
+    formatter.writeToStream(writeStream);
+  });
+};
+
+const saveZippedSessionsElectron = (exportedSessionIds, sessions, installedProtocols) => {
+  // eslint-disable-next-line global-require
+  const electron = require('electron');
+  const fs = window.require('fs');
+  const { dialog, shell } = window.require('electron').remote;
+  const tempPath = (electron.app || electron.remote.app).getPath('temp');
+
+  const defaultFileName = 'networkCanvasExport.zip';
+
+  return new Promise((resolve, reject) => {
```

Comment:
I'd opt for resolving this earlier and leaning on the promises more for readability e.g.:

```
// The dialog method only needs to capture the filename, then it can resolve
const saveDialog = () => 
  new Promise((resolve, reject) => {
    dialog.showSaveDialog({
      filters: [{ name: 'zip', extensions: ['zip'] }],
      defaultPath: defaultFileName,
    // eslint-disable-next-line consistent-return
    }, (filename) => {
      if (filename === undefined) resolve({ cancelled: true });

      resolve({ cancelled: false, filename });
    });
  });

// Then we can start to compose the promise from parts with understandable names:
const exportFiles = (exportedSessionIds, installedProtocols, tempPath) => {
  const promisedExports = exportedSessionIds.map((sessionId) => {
    const session = sessions[sessionId];
    const sessionCodebook = installedProtocols[session.protocolUID].codebook;
    const dataFileName = path.join(tempPath, `${session.caseId}_${sessionId}.graphml`);

    // export one file at a time
    return exportElectronFile(fs, session, sessionCodebook, dataFileName);
  });

  return Promise.all(promisedExports);
};

const exportArchive = filename =>
  exportFiles()
    .then(sourceFiles =>
      // by nesting when you need to keep things in context, rather than breaking out into let
      // variables we can make these promises really composable...
      archive(sourceFiles, filename))
        .then(() => sourceFiles.map(dataFileName => tryUnlink(dataFileName, fs)),
    );


// The finally function is much easier to follow (and hopefully the constituent parts too!):
saveDialog()
  .then(({ cancelled, filename }) => {
    if (cancelled) { return; }

    return exportArchive(filename)
      .then(() => shell.showItemInFolder(filename));
  });
```

Reply:
Good call, thanks!